Wasm-container/README.md

```diff
@@ -0,0 +1,509 @@
+
+# An experiment with creating Wasm UDxes
+
+In the `examples/UDx` directory you will find prototype UDxes written
+in C-compiled-to-Wasm and Rust-compiled-to-Wasm.
+
+There is a `Makefile` in that directory with the following `make`
+targets:
+
+- cWasmUDxlib --- A UDx which loads a `sum` function from `sum.c.wasm`
+- rustWasmUDxlib --- A UDx which loads a `sum` function from `sum.rs.wasm`
+- nonWasmUDxlib --- A UDx which does the `sum` calculation directly
+
+These UDxes consist largely of boilerplate.  There are different files
+to make it easy to compare the performance of the three UDx
+implementations against one another in the same Vertica instance.
+
+### An example Rust UDx
+
+Here is a sample Rust implementation of a simple `sum` function:
+
+```rust
+#[no_mangle]
+pub extern "C" fn sum(a: u32, b: u32) -> u32 {
+    return a + b;
+}
+```
+
+The corresponding C++ boilerplate is in
+`examples/UDx/rustWasmUDx.cpp`.  For an explanation of the
+boilerplate, see [Extending
+Vertica](https://www.vertica.com/docs/latest/HTML/Content/Authoring/ExtendingVertica/ExtendingVertica.htm).
+
+What we're doing here is telling Vertica that we've produced a C++
+UDx.  That the C++ UDx happens to load and interpret another file to
+do the bulk of its computation isn't important to Vertica.
+
+As with other C++ UDxes, one creates a factory class (here called
+`rustWasmUDx_sumFactory`) and a function class (here called `rustWasmUDx_sum`).  In this case, the factory class exists
+primarily to tell Vertica the function prototype for the
+`rustWasmUDx_sum` function --- that it takes two integer arguments
+(`argTypes.addInt()`) and returns an integer `returnType.addInt()`).
+
+Function classes can be complicated, but for our simple function, it
+has three methods:
+
+- `setup`: performed once when Vertica is informed of the function.
+  In this case we call `udx_get_wasm_state()` to get a `wasm_state` to
+  hold information about the Wasm C API, and to attach that state to
+  the Wasm bytecode file with `udx_setup()`.  Note that `udx_setup`
+  binds this `wasm_state` to the "sum" function.
+
+- `destroy`: called to deallocate the `wasm_state` data structures.
+
+- `processBlock`: called once for each row of the table being
+  processed.  In this case, the function reads the arguments (using
+  `argReader.getIntRef`, then calls the function with the two
+  arguments (using `udx_call_func_2i_1i`, since this is a
+  2-int-argument, 1-int-return function, and then uses
+  `resWriter.setInt` to return the result.  `resWriter.next` is used
+  to indicate that we've written all the results for this row, and
+  `argReader.next` is used to read the next row.
+
+# Loading and executing the UDx
+
+## Starting a test Vertica using the container
+
```

Comment:
Tutorials are usually written as either second person ("you can do X" or imperative) or with 'we" language (that's more in academia I think, but I don't think we have strong style guidelines for our GitHub readmes so I don't care), but "I" language feels a little odd in this context.